### PR TITLE
Changes for OpenSAML3 migration

### PR DESCRIPTION
**Pull request by deshankoswatte**
### Proposed changes in this pull request
Changes for OpenSAML3 migration

### When should this PR be merged
Not specified yet

### Prerequisites
- wso2/orbit#395 - Orbit for OpenSAML3
- wso2/carbon-identity-framework#2473 - Migration from OpenSAML 2.x to 3.x
- wso2-extensions/identity-metadata-saml2#46 - Migration from OpenSAML 2.x to 3.x
- https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/256 - Migration from OpenSAML 2.x to 3.x

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 1]
```diff
@@ -546,9 +541,20 @@ protected void validateSignatureAgainstIdpCertificate(Assertion assertion, Strin
         X509Certificate x509Certificate = getIdpCertificate(tenantDomain, identityProvider);
         try {
             X509Credential x509Credential = new X509CredentialImpl(x509Certificate);
-            SignatureValidator signatureValidator = new SignatureValidator(x509Credential);
-            signatureValidator.validate(assertion.getSignature());
-        } catch (ValidationException e) {
+
+            // The process mentioned below is done because OpenSAML3 does not support OSGi refer
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Add the stack overflow link too

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 2]
```diff
@@ -124,17 +125,11 @@ public void init() throws IdentityOAuth2Exception {
 
         super.init();
 
-        Thread thread = Thread.currentThread();
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Check if this process should also persist with the changes

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 3]
```diff
@@ -124,17 +125,11 @@ public void init() throws IdentityOAuth2Exception {
...
-        ClassLoader loader = thread.getContextClassLoader();
-        thread.setContextClassLoader(this.getClass().getClassLoader());
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Check if this process should also persist with the changes

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ public X509Certificate getEntityCertificate() {
     }
 
     @Override
-    public CredentialContextSet getCredentalContextSet() {
+    public CredentialContextSet getCredentialContextSet() {
         // This method is not implemented
         return null;
     }
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java]: Change the method to:

```
/***
 * get a credential context set
 * @return Method not supported so will return null
 */
public CredentialContextSet getCredentialContextSet() {
      return null;
}
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java [hunk 5]
```diff
@@ -135,6 +135,6 @@ public SecretKey getSecretKey() {
     @Override
     public UsageType getUsageType() {
         // This method is not implemented
```
Comment (deshankoswatte) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java]: Remove TODO and say: method not supported in turn will return a default value

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java (hunk 1 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Added

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java (hunk 4 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java]: Changed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java (hunk 5 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/X509CredentialImpl.java]: Fixed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java (hunk 3 above)
Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Decided to keep them as it is

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java [hunk 6]
```diff
@@ -546,9 +549,23 @@ protected void validateSignatureAgainstIdpCertificate(Assertion assertion, Strin
...
-        } catch (ValidationException e) {
+
+            /*
+              The process mentioned below is done because OpenSAML3 does not support OSGi refer
+              https://shibboleth.1660669.n2.nabble.com/Null-Pointer-Exception-from-UnmarshallerFactory-while-migrating-from-OpenSAML2-x-to-OpenSAML3-x-td7643903.html
+              and https://stackoverflow.com/questions/37948303/opensaml3-resource-not-found-default-config-xml-in-osgi-container
+            */
+
+            Thread thread = Thread.currentThread();
+            ClassLoader loader = thread.getContextClassLoader();
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: rename `loader` to `originalClassLoader`

Comment (deshankoswatte) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java]: Fixed

**Pull request merged**